crates/schema/src/type_for_generate.rs

```diff
@@ -0,0 +1,516 @@
+    /// This is necessary to disambiguate between a use of the unit type, and a reference to a type declaration of a product type with no elements.
+    known_definitions: HashSet<AlgebraicTypeRef>,
+    /// Interning data structure, no semantic meaning.
+    uses: HashSet<Arc<AlgebraicTypeUse>>,
+    /// AlgebraicTypeRefs that point to uses.
+    known_uses: HashMap<AlgebraicTypeRef, AlgebraicTypeUse>,
+    /// Used for cycle detection.
+    currently_touching: HashSet<AlgebraicTypeRef>,
+}
+
+impl TypespaceForGenerateBuilder<'_> {
+    /// Finish building the `TypespaceForGenerate`.
+    pub fn finish(self) -> TypespaceForGenerate {
+        self.result
+    }
+
+    /// Add a *type use* to the typespace.
+    /// This is a *use* of a type, not a definition.
+    /// The input must satisfy `AlgebraicType::is_valid_for_client_type_use`.
+    /// The input must resolve successfully.
+    /// The input must not contain any cyclic definitions.
+    pub fn add_use(&mut self, ty: &AlgebraicType) -> Result<AlgebraicTypeUse> {
+        if ty.is_address() {
+            Ok(AlgebraicTypeUse::Address)
+        } else if ty.is_identity() {
+            Ok(AlgebraicTypeUse::Identity)
+        } else if ty.is_unit() {
+            Ok(AlgebraicTypeUse::Unit)
+        } else if ty.is_never() {
+            Ok(AlgebraicTypeUse::Never)
+        } else if let Some(elem_ty) = ty.as_option() {
+            let elem_ty = self.add_use(elem_ty)?;
+            let interned = self.intern_use(elem_ty);
+            Ok(AlgebraicTypeUse::Option(interned))
+        } else if ty.is_schedule_at() {
+            Ok(AlgebraicTypeUse::ScheduleAt)
+        } else {
+            match ty {
+                AlgebraicType::Ref(ref_) => {
+                    // This is the only seriously complicated case, which has to deal with cycle detection.
+                    let ref_ = *ref_;
+
+                    if self.known_definitions.contains(&ref_) {
+                        self.add_definition(ref_)?;
+                        // The ref points to a known definition, so it must be a use.
+                        Ok(AlgebraicTypeUse::Ref(ref_))
+                    } else if let Some(use_) = self.known_uses.get(&ref_) {
+                        // The ref is to a use which we have already seen.
+                        Ok(use_.clone())
+                    } else {
+                        // We haven't processed it yet. It's either a ref to a valid use, or invalid.
+                        let def = self
+                            .typespace
+                            .get(ref_)
+                            .ok_or(TypeRefError::InvalidTypeRef(ref_))
+                            .and_then(|def| {
+                                if def == &AlgebraicType::Ref(ref_) {
+                                    // Self-reference.
+                                    Err(TypeRefError::RecursiveTypeRef(ref_))
```

Comment:
So this now prohibits recursive types altogether? Can we please keep them?

I currently can autogenerate C# AlgebraicType structure - which is recursive - with all its children types directly from Rust, and I'd like to make use of it after landing ModuleDef autogen in #1670.

If we error out on recursive types, then AlgebraicType as well as anything that depends on it - which includes ModuleDef - will error out and won't codegen anymore.

I think instead of an error we should have `AlgebraicTypeUse::Ref { ref_, is_recursive }` variant so that langs that need to explicitly box the reference (only Rust among langs we currently support), have boolean tag to know that they need to do so, whereas others that already keep all the objects on the heap, can deal with it as a normal reference.

Reply:
Ah, I thought we didn't support recursive types in modules though. Does the code generation for that pass through this, like are you using the module codegen to generate those? I'm happy to add that tag if so.

Reply:
> Does the code generation for that pass through this, like are you using the module codegen to generate those?

Yeah, since we don't have any other codegen, we use regular `spacetime generate` for ClientApi and for this new ModuleDef and then delete files we're not interested in heh.

Someday we might clean it up, but I imagine the basic codegen pipeline is still going to be shared between modules and "just types".

Reply:
Cool, I'll fix that then.
Ugh now I need to actually handle cycles though. ;-;

Reply:
Luckily, you use `Arc` so it should be as easy as downgrading it all to `sync::Weak`... maybe? hopefully? 😅

Reply:
The cycles all pass through `AlgebraicTypeRef` so no need for that even.